### PR TITLE
Allow manual memory reset in AttentionMechanism

### PR DESCRIPTION
**Pull request by kazemnejad**
Add support for manual memory reset by removing the `_memory_initialized` cache on `mechanism(..., setup_memory=True)` call

Fixed: #535

### CODE REVIEW

File: tensorflow_addons/seq2seq/attention_wrapper_test.py [hunk 1]
```diff
@@ -209,6 +209,57 @@ def test_masking(self):
...
+        ("bahdanau_monotonic", wrapper.BahdanauMonotonicAttention),
+    )
+    def test_memory_re_setup(self, attention_cls):
+        class MyModel(tf.keras.models.Model):
+            def __init__(self):
+                super(MyModel, self).__init__()
+                self.emb = tf.keras.layers.Embedding(
+                    vocab, embedding_dim, mask_zero=True)
+                self.encoder = tf.keras.layers.LSTM(
+                    test_class_self.memory_size, return_sequences=True)
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/attention_wrapper_test.py]: you can pass those value in as init param, which is standard practice to build a custom model.

File: tensorflow_addons/seq2seq/attention_wrapper_test.py [hunk 2]
```diff
@@ -209,6 +209,57 @@ def test_masking(self):
...
+                enc_input = self.emb(enc_input)
+                enc_output = self.encoder(enc_input, mask=mask)
+                # To ensure manual resetting also works in the graph mode,
+                # we call the attention mechanism twice.
+                self.attn_mch(enc_output, mask=mask, setup_memory=True)
+                self.attn_mch(enc_output, mask=mask, setup_memory=True)
+                score = self.attn_mch([query, state])
+                return score
+
+        test_class_self = self
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/attention_wrapper_test.py]: you can ignore this now.

**Comment by guillaumekln** [comment]
Thanks @kazemnejad!

**Comment by qlzh727** [comment]
Indeed. Thanks @kazemnejad for the change.

**Pull request merged**